datastream-kafka/src/main/java/com/linkedin/datastream/kafka/KafkaProducerWrapper.java

```diff
@@ -254,67 +298,103 @@ void send(DatastreamTask task, ProducerRecord<K, V> producerRecord, Callback onC
       _kafkaProducer = null;
-    }
 
-    // This may be called from the send callback. The callbacks are called from the sender thread, and must complete
-    // quickly to avoid delaying/blocking the sender thread. Thus schedule the actual producer.close() on a separate
-    // thread
-    if (producer != null) {
+      // This may be called from the send callback. The callbacks are called from the sender thread, and must complete
+      // quickly to avoid delaying/blocking the sender thread. Thus schedule the actual producer.close() on a separate
+      // thread
       _producerCloseExecutorService.submit(() -> {
         _log.info("KafkaProducerWrapper: Closing the Kafka Producer");
-        producer.close(CLOSE_TIMEOUT_MS, TimeUnit.MILLISECONDS);
-        NUM_PRODUCERS.decrementAndGet();
-        _log.info("KafkaProducerWrapper: Kafka Producer is closed");
+        try {
+          producer.close(CLOSE_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+          NUM_PRODUCERS.decrementAndGet();
+          _log.info("KafkaProducerWrapper: Kafka Producer is closed");
+        } finally {
+          markProducerCloseComplete();
+        }
       });
+    } finally {
+      _producerLock.unlock();
     }
   }
 
-  private DatastreamRuntimeException generateSendFailure(Exception exception) {
+  private void markProducerCloseComplete() {
+    try {
+      _producerLock.lock();
+      _closeInProgress = false;
+      _waitOnProducerClose.signalAll();
+    } finally {
+      _producerLock.unlock();
+    }
+  }
+
+  private DatastreamRuntimeException generateSendFailure(Exception exception, DatastreamTask task) {
     _dynamicMetricsManager.createOrUpdateMeter(_metricsNamesPrefix, AGGREGATE, PRODUCER_ERROR, 1);
     if (exception instanceof IllegalStateException) {
       _log.warn("Send failed transiently with exception: ", exception);
       return new DatastreamTransientException(exception);
     } else {
       _log.warn("Send failed with a non-transient exception. Shutting down producer, exception: ", exception);
-      shutdownProducer();
+      if (_tasks.contains(task)) {
+        shutdownProducer();
+      }
       return new DatastreamRuntimeException(exception);
     }
   }
 
   void flush() {
-    synchronized (_producerLock) {
+    Producer<K, V> producer;
+    try {
+      _producerLock.lock();
+      producer = _kafkaProducer;
+    } finally {
+      _producerLock.unlock();
+    }
+
+    try {
+      if (producer != null) {
+          producer.flush(_producerFlushTimeoutMs, TimeUnit.MILLISECONDS);
+      }
+    } catch (InterruptException | TimeoutException e) {
+      // The KafkaProducer object should not be reused on an interrupted flush
       try {
-        if (_kafkaProducer != null) {
-          _kafkaProducer.flush(_producerFlushTimeoutMs, TimeUnit.MILLISECONDS);
+        _producerLock.lock();
+        if (producer == _kafkaProducer) {
+          _log.warn("Kafka producer flush interrupted/timed out, closing producer {}.", producer);
```

Comment:
Is the interruption/timeout typical? Wondering about the use of log.warn which are not too noticeable in logs.. wondering if log.error might be better, but I know it raises false alarms in thinking something is wrong when it is being handled here fine.

Reply:
This will be useful in identifying the producer stuck on flush issues. Yes, you are correct. The reason to not add it as ERROR  is to avoid false alarm. This is a self-recovering scenario which should not need extra attention. So, warns will be sufficient to  identify if this issue happened.